### ISSUE TITLE
Caching static assets

### ISSUE THREAD

**Issue by ojeytonwilliams** [freeCodeCamp/freeCodeCamp#42107]
### The problem

Pagespeed insights gives https://www.freecodecamp.org/ a relatively poor score (50-60 for mobiles and 80-90 for desktops) indicating that there is room for improvement.

One of the issues is that we do not cache our static assets for very long:
![image](https://user-images.githubusercontent.com/15801806/117874584-6fcb3800-b2a1-11eb-819a-7aca9558fb17.png)

while this does not directly impact our score it *does* directly impact our users.  They're wasting time and bandwidth downloading the same assets multiple times.

### The cause of the problem

We rely on Cloudflare default browser cache, with a TTL of 4 hours.  What they do is simply add `cache-control: max-age=14400` to whatever they serve.

### The (proposed) solution

Since webpack (both in our `npm run build-workers` and in Gatsby) adds content hashes to the name of each asset we can store them indefinitely.  When the client is updated, it will request new versions of the assets, since they will have new content hashes.  This should be true for everything except `bootstrap.css` and `sass.sync.js`.  Every other asset can then be cached with `Cache-Control: public, max-age=604800, immutable` as discussed on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#caching_static_assets).

Since this is an aggressive caching strategy, we should make sure that all assets which do *not* include content hashes are accounted for.

#### Setting it via Cloudflare

If `Cache-Control: max-age 31536000` is enough, then we can just let Cloudflare handle it.  I'm not sure `public` or `immutable` is strictly necessary.  That's live right now on https://www.freecodecamp.dev/.

#### Controlling the headers ourselves

If fine-tuning is required, we can set them and Cloudflare will respect the headers we set.

#### 404s

If the asset is missing, Cloudflare will cache it for 3 minutes, [by default](https://support.cloudflare.com/hc/en-us/articles/200172516-Understanding-Cloudflare-s-CDN). If we set `Cache-Control no-cache` it will not be cached by Cloudflare.  Ideally we should set this to minimise disruption for users when we deploy.

**Comment by raisedadead** [comment]
Thanks for the detailed spec and what are the limitations with our current stack, preventing us from implementing this?

**Comment by ojeytonwilliams** [comment]
The only issue I ran into was how to add `Cache-Control` headers to 404s.  We can do that easily with `serve`, as they show [here](https://github.com/vercel/serve-handler#headers-array), but I haven't been able to get NGINX to do that.

`proxy_intercept_errors` might the only change required to get NGINX to play ball, but I've not tested it yet.

**Comment by raisedadead** [comment]
OK - Thanks for the additional details. 

Going through the details I am failing to understand the "The (proposed) solution" part. Are these:

- Setting it via Cloudflare
- Controlling the headers ourselves
- 404s

Three different approaches OR overlapping approaches?

**Comment by ojeytonwilliams** [comment]
So, the problem is that serve is the only one I've been able use for both 200s and 404s.  Cloudflare and NGINX alone won't cut it:

Cloudflare can set `max-age` for 200s.
NGINX can set whatever we want for 200s (404s I'm not sure about, I'm still experimenting)

So, Cloudflare alone can't do everything (at least not without involving workers).  If we set `max-age` via Cloudflare we'd need something else to set `no-cache` for 404s.

NGINX _might_ be enough on it's own, I'll keep you posted.

**Comment by raisedadead** [comment]
My honest recommendation is not chasing an implementation just yet, to save you time and efforts. Could you [confirm my previous query](https://github.com/freeCodeCamp/freeCodeCamp/issues/42107#issuecomment-839999357)?

Finally, the OP is slightly vague on the application that it is targeting. I am assuming you are talking about learn all this while.

**Comment by ojeytonwilliams** [comment]
Yes, this is about learn. As I said initially, the pagespeed insights report for https://www.freecodecamp.org/ is pretty bad.  /learn is even worse (it scores around 40 for mobile)

> Three different approaches OR overlapping approaches?

Neither NGINX nor Cloudflare alone solves the problem, so they would need to be used in conjunction with serve.  So, there are three possibilities: serve + Cloudflare, serve + NGINX or serve alone.

**Comment by raisedadead** [comment]
Thanks.

I am going to clean up the fragmented information in the thread here and attempt to summarise the issue (before beginning to attack it with any implementation details - that's the easy bit):

1. The landing page (let's call it what it is) and the coding platform (at `/learn`) are the same. They are served from the same origin servers with varying JS, CSS, and other assets along with the primary entry points that are the corresponding `.html` files for a path/page.

   This is primarily a SPA mounted at the root of the `freecodecamp.org` domain. So aptly, let's call this the root app (we refer to this as `client` in the codebase).

2. Then we have the forum app, the coderadio app, the news app, and the localized apps (with their supporting forum and news, etc.). 

   We can revisit the pagespeed targets for them once we have tackled the root application.

Now that we are focusing on the root app. **Please note a good UX is what we are after, to be corroborated by pagespeed insights and not the other way round**.

Here is a decent strategy for a good UX (in simple language user stories).

1. Identify the entry points for the root app (off the top of my head, these are all the `.html` files)
2. Ensure that the entry points are always fetched from the origin server.
   - The origin server, in this case, is a web server (it can be any web server, express, NGINX, serve, doesn't matter).
   - Note: I am deliberately using the term "entry point" and not 404s, etc., because that is what a browser will parse before getting to other resources within the entry points.
   - The headers are to be set by the webserver (which you correctly identified as `serve` for the current implementation - we will not have that soon, I will explain why later)
3. Ensure that the "chain" of proxies in front of the webserver respects the headers and passes them forward. For our stack, this chain looks like so `Origin Web Server: NGINX Proxy --> Azure Load Balancer --> Cloudflare Edge Nodes --> Browser`.
4. Adding a `no-cache` & `max-age` or just a header for that matter are not the only things that will help a good UX. They will only check off one metric in that "diagnostics" list in the OP.
5. We would also need to implement additional practices like using `ETags`, and actually serving the assets forever (more like a year) within the entry points.
6. We would need to do a better fetching of updates to the SPA (See https://github.com/freeCodeCamp/freeCodeCamp/issues/39612) - this is one of the core reason for broken blank pages from time to time (what I believe you have been referring to as 404s?? IDK).
7. Finally, we need to decouple the responsibilities in the **"HOW"** part of the root apps delivery chain. At the minute, the business logic is spread across `serve`, `NGINX`, and Cloudflare.

Let me get back to you with an actionable list of tasks for these user stories.

**Comment by ojeytonwilliams** [comment]
I agree with argument overall, so I'll only respond where I disagree/want to clarify.

> Adding a no-cache & max-age or just a header for that matter are not the only things that will help a good UX. They will only check off one metric in that "diagnostics" list in the OP.

I agree with this.  This issue has limited scope, though, I focused on the headers as one way we can improve.  The vast majority of the problems we have stem from the size of the js bundles. Something @ahmadabdolsaheb and I have been addressing, but it's an ongoing project.

> We would also need to implement additional practices like using ETags

ETags are only really useful for assets that change over time, but whose name is fixed.  For the assets we serve we can generate new contentHashes whenever their content changes and then cache them for a year.

> and actually serving the assets forever (more like a year) within the entry points.

Yep, we shouldn't rely exclusively on Cloudflare for this.  I believe you were planning to keep old assets as S3 buckets, which makes sense to me.

> We would need to do a better fetching of updates to the SPA (See #39612) - this is one of the core reason for broken blank pages from time to time (what I believe you have been referring to as 404s?? IDK).

Let's say a user reloads any page after deployment, they should be fine since we are not caching entry points.  All the new assets will be available, as will the old ones (in Cloudflare's cache).

Let's say a user navigates within the SPA *without* reloading.  Again, I *think* they're fine, because they should be exclusively using old assets (fetched when they requested the old entry point), though I'm not 100% certain.

Let's say a user reloads any page *during* deployment.  Since we have multiple origins and a proxy chain, they could get a new entry point which then requests an asset and may get routed to a different origin.  This asset is not guaranteed to be present, hence 404s can result.

We can mitigate this in two ways: sync up client releases (this is already decent, since we release in parallel, but it may be possible to fine tune) and prevent the caching of 404 results.  Then if a user does not get an asset they need, reloading should solve it.

> Finally, we need to decouple the responsibilities in the "HOW" part of the root apps delivery chain. At the minute, the business logic is spread across serve, NGINX, and Cloudflare.

Yep, it's not great.  If we're ripping out serve, I'd prefer to handle Cache-Control headers exclusively in the NGINX layer (if possible, but I'm not sure how).

**Comment by raisedadead** [comment]
> ETags are only really useful for assets that change over time, but whose name is fixed.

Correct - All of our entry points (`index.html`s, `404.html`s, etc.) in the SPA are exactly that. 

We can either not cache them at all, which is great, OR we can smartly use ETags to speed up delivery for the entry points. Cloudflare will do the last mile delivery for us from its edge nodes.

> we release in parallel, but it may be possible to fine tune

The goal is to remove the ambiguity by the browser on the assets that are to be pulled. Sometimes, it requests an asset that is not readily available (to be explicit, not reachable via the origin webserver - asset could be on the disk already).

There is a way to achieve this: have one & only one origin web server. That way, regardless of when the browser requests the entrypoint, it will also have access to the entry point's needed assets.

I know Gatsby (our SPA engine) has a way to achieve this (namely, using a CDN path). I am exploring an economical way to do this. 

The underlying idea is:

**Storage** (S3) + **Dynamic Mapping** to a release (Minio) +  **Web Server** for the HTTP handling (NGINX??) is on the table.

However, as I've said before, I will need to get back to you on this.

> If we're ripping out serve, I'd prefer to handle Cache-Control headers exclusively in the NGINX layer

This is a gray area at this time. We want to have as few hops as possible. I will keep you posted. 

I am not sure why it is difficult to set blanket headers on NGINX for all requests that it proxies to an origin, because we know the MIME types of the entry points that we want to do a `no-cache` policy on.

**Comment by ojeytonwilliams** [comment]
> can smartly use ETags to speed up delivery for the entry points. Cloudflare will do the last mile delivery for us from its edge nodes.

Ah, nice, I'd not thought about that, but it could be a big improvement.  Are you looking into it or shall I?

> Storage (S3) + Dynamic Mapping to a release (Minio) + Web Server for the HTTP handling (NGINX??).
> 
> However, as I've said before, I will need to get back to you on this.

Sounds good.  Let me know if I can help.

> I am not sure why it is difficult to set blanket headers on NGINX for all requests that it proxies to an origin, because we know the MIME types of the entry points that we want to do a no-cache policy on.

It probably isn't, but I'm not the greatest at configuring NGINX.  I'll see what I can find out.

**Comment by raisedadead** [comment]
> Ah, nice, I'd not thought about that, but it could be a big improvement. Are you looking into it or shall I?

I am not looking into any perf-related issues at all. Please feel free to explore.

**Comment by shadmankhan** [comment]
Hi guys, I understand the importance of caching static assets. I have a suggestion, that we can add this caching process in the service-worker at the client-side. This also has various pros, that at this moment FCC doesn't have PWA (Progressive web app) support. We can bring the PWA feature as well as caching to the client-side. To achieve this, there are several ways, one of them is to use Google's workbox.

I hope I am able to explain the idea.
Please let me know your feedback. :)
